find_duplicates.py

```diff
@@ -0,0 +1,249 @@
+#!/usr/bin/env python3
+#
+# Detect and delete duplicated flows in routers defined in OVN.
+# Steps:
+#
+# - Get the output of 'ovn-sbctl lflow-list <datapath>', where <datapath> is a
+#   router's name (e.g. zuul-tests_router).
+# - Group the list of logical flows dropping the "action=..." tail.
+# - For each flow of the previous list that has 2 or more ocurrences the
+#   associated port is deleted. To identify the right port the following steps
+#   are taken:
+#   + Get the output of 'ovn-nbctl list logical-switch-port' where each section
+#     (separated by an empty line) is a port.
+#   + Get the MAC and IP addresess from the flow that wants to be deleted.
+#   + Search in the list of ports the port that has the MAC and IP addresses in
+#     the `addresses` field AND `up` is false.
+#
+from __future__ import annotations
+
+import argparse
+import collections
+import functools
+import json
+import os
+import re
+import subprocess
+import sys
+import uuid
+
+
+__doc__ = 'Detect and delete duplicated logical flows.'
+
+RE_LRP = re.compile('^[ ]+table=\d+\(lr_in_arp_resolve  \), priority=\d+  , '
+                    'match=\(outport == "lrp-.*')
+RE_PORT_UUID = re.compile('.*\(outport == "lrp-(.*)" ')
+RE_IP_ADDR = re.compile('.* reg0 == (.*)\), action=\(eth.dst = ([a-f0-9:]{17});')
+
+
+class PortNotFound(Exception):
+    pass
+
+
+class ManyPortsFound(Exception):
+    pass
+
+
+def setup_options() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(description=__doc__)
+    parser.add_argument('--datapath', dest='datapath', metavar='NAME',
+                        required=True,
+                        help='Datapath')
+    parser.add_argument('--delete', dest='delete', action='store_true',
+                        help='Delete duplicates found')
+    parser.add_argument('--dry-run', dest='dry_run', action='store_true',
+                        help=("Don't perform the operation, just print what "
+                              "deletions would have been made."))
+    return parser.parse_args()
+
+
+def get_ovn_ips(name: str) -> str:
+    """Get ONV NB or SB IP addresses to connect to."""
+    assert name in ["ovnnb", "ovnsb"]
+
+    with open('/etc/ovn/ovn-northd-db-params.conf', 'r') as f:
```

Comment:
Relying on presence of this file will require the user to run this script from one of the `ovn-central` units. In theory they could run it from anywhere granted they have certs.

OTOH, the charm will set up UFW rules restricting access, so running it there might be the best choice anyway.

Perhaps mention where it is expected to be run in the docs in the top of the file?

Reply:
yes, probably the best will be to check file's presence early and print a message indicating where this program is expected to run on. I will do the change.